src/engine/engine.cpp

```diff
@@ -303,4 +303,39 @@ OutputParameters Engine::execute_model(const std::vector<Sequence*>& batch) {
   return first_output;
 }
 
+// TODO
+OutputParameters Engine::validate(const std::vector<Sequence*>& batch) {
+  torch::Tensor flatten_token_ids;
+  torch::Tensor flatten_positions;
+  torch::Tensor seq_idxes;
+  InputParameters input_params;
+  SamplingParameters sampling_params;
+
+  Utils::prepare_validate_inputs(batch,
+                                 FLAGS_block_size,
+                                 &flatten_token_ids,
+                                 &flatten_positions,
+                                 &seq_idxes,
+                                 &input_params,
+                                 &sampling_params);
+  if (workers_.size() == 1) {
+    auto output = workers_[0]->validate(
+        flatten_token_ids, flatten_positions, input_params, sampling_params);
+    output.index_select(seq_idxes);
+    return output;
+  }
+
+  std::vector<folly::SemiFuture<OutputParameters>> futures;
```

Comment:
nit: in my understanding, all workers should return same logits from last step if they are initialized with same seed. If it is the case, sounds validattion can be done only on one worker.

Reply:
good catch here.